File: 백준/Silver/15652. N과 M （4）/N과 M （4）.java
import java.io.*;
import java.util.*;
public class Main{
    static boolean[] visit;
    static int[] arr;
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        visit = new boolean[N + 1];
        arr = new int[M];
        solution(N, M, 0);
    }

    private static void solution(int n, int m, int cnt) throws IOException {
        if (cnt == m) {
            for (int num : arr) {
                System.out.print(num + " ");
            }
            System.out.println();
        } else {
            for (int i = 1; i <= n; i++) {
                if(cnt > 0 && i < arr[cnt-1]) {
                    continue;
                }
                arr[cnt] = i;
                solution(n, m, cnt + 1);
            }
        }
    }

}

File: 백준/Silver/15652. N과 M （4）/N과 M （4）.py
n, m = map(int, input().split())
arr = [0] * m
isUsed = [0] * (n + 1)
def func(cnt, s):
    if cnt == m:
        [print(i, end=' ') for i in arr]
        print()
        return
    for i in range(s, n+1):
        arr[cnt] = i
        func(cnt + 1, i)
func(0, 1)